### PR TITLE
Generate Async Overloads for All Public Functions

### LINKED ISSUE #332: Add async/await support
**Description**
Since server-side Swift projects like Vapor now support async/await APIs, it would nice if meilisearch-swift also added this.

**Basic example**
As opposed to the completion parameter style currently used, an async/await call would look like so:

    let result = await index.updateDocuments(documents: [/** data **/], primaryKey: "id")
    switch result {
    case .success:
        // ....
    case .failure(let error):
        // ....
    }

**Proposed solution**
I have already created extensions in my local project to support some of the methods that I use. For example `updateDocuments()` is covered by this implementation:

    public func updateDocuments<T>(
      documents: [T],
      encoder: JSONEncoder? = nil,
      primaryKey: String? = nil) async -> Result<TaskInfo, Swift.Error> where T: Encodable {
          return await withCheckedContinuation { continuation in
              self.updateDocuments(documents: documents, encoder: encoder, primaryKey: primaryKey) { res in
                  continuation.resume(returning: res)
              }
          }
    }

Proposal is to add these bridging methods for _all_ methods on `Indexes` in perhaps an `Indexes+async.swift` extension. Additionally unit tests will need to be added to cover async calls.

### PR DESCRIPTION
**Pull request by Sherlouk**
# Pull Request

## Related issue
Fixes #332 

This is NOT a breaking change.

## What does this PR do?
- Adds async/await overloads to all public interfaces.

The contents of these files has been automatically generated using [Sourcery](https://github.com/krzysztofzablocki/Sourcery) using [this template](https://gist.github.com/Sherlouk/fa3e49e24f9be7232c642b945177b05e) I made. It does not include documentation, I feel like this is an okay compromise at this point in time and is something that can be incremented on in the future, potentially when we move over entirely. Async/Await functions sit in Xcode along with the normal closure based methods and so is easy to see the documentation inline anyway.

The template has been shared to hopefully make it easier to update when new APIs are added in the future, keeping maintenance down.

With the previous PR (#410) I called the underlying implementation directly, this leads to unnecessary duplication with other functions. Instead I call the non-async function guaranteeing consistent functionality, and no unnecessary duplication. I also feel like this avoids the need for excessive and additional test coverage, so long as the build compiles then the implementation is working as expected and tests give us no extra confidence that it's working as expected. We'll keep the one for search to demonstrate.

@aronbudinszky Are you happy with this approach too?

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

File: Sources/MeiliSearch/Async/Client+async.swift
Comment (Sherlouk) [inline Sources/MeiliSearch/Async/Client+async.swift:190]: This implementation is very slightly different to all others - this is because the underlying implementation does not return a result, but rather returns `false` if an error is identified.

As such we use a checked continuation (no throwing), and forward the bool response verbatim.

Works the same, just no throwing.

**Review by curquiza — APPROVED** [review]
I'm ok with this PR, thank you!

Let's wait for the @aronbudinszky feedback before merging.
If we don't get any answer, we will merge it.

File: Sources/MeiliSearch/Async/Client+async.swift
Comment (aronbudinszky) [reply Sources/MeiliSearch/Async/Client+async.swift:190]: Nice.

However this does point to an important omission that I would say blocks this PR currently - the lack of tests.

I think one way to test things might be to check that the async results are equivalent to the non-async ones - this would avoid duplicating tests but still ensure our assumptions about success, failure, and the compatibility with the old-style api.

(If you have time cool, if not I can probs take a look at it as well in the next few days.)

**Review by aronbudinszky — CHANGES_REQUESTED** [review]

**Comment by aronbudinszky** [comment]
Sorry somehow missed your comment about tests above - though I _kinda_ agree that parity is _theoretically_ assured, I still think that tests are important because going to async can sometimes cause weird issues and/or there can be _some_ exceptions to how things are implemented exactly.

Also in case there are any test coverage requirements (which probs there should be) this would help with those stats... :)

**Comment by aronbudinszky** [comment]
@curquiza Just to confirm, this is *not* a breaking change, so if we choose to add tests in an upcoming PR I'm also happy to merge this in.

**Comment by aronbudinszky** [comment]
(On a side note as I see there are [plans](https://github.com/meilisearch/meilisearch-swift/issues/405) to not allow untested code to be merged.)

**Comment by curquiza** [comment]
Thanks @aronbudinszky for your reactivity.
About tests you are right, and I would rather we have the tests in the same PR than the addition, if possible 😊

**Comment by Sherlouk** [comment]
Spent a little bit of time looking into the tests, and so wanted to see what folks thought a good balance would be for now:

1. Duplicate all tests so every functionality has an async/await test and a closure based test
2. Replace all closure based tests with async/await tests (under the basis that every async implementation calls the closure based one so in terms of test coverage it's identical and all integrations are touched)
3. Create minified tests to async/await to simply assert a request to the correct API URL path but no assertions on response type (which is covered by existing closure tests).

I'm wary of adding too much duplication and maintenance overhead with this PR, especially until I get to rewriting some of the unit and integration tests to reduce prose. 

My opinion would be continuing with number two for now, moving us in the strategic direction of defaulting to async/await, reducing complexity in tests (with use of expectations), and by extension improving performance of tests. It still guarantees a 100% test coverage for public APIs.

**Comment by Sherlouk** [comment]
I've pushed up a couple examples of my proposal to aide in discussion, the new tests are running about 80% faster locally. 

I also believe them to be far more readable, though this will be especially prevalent when I get to the more complex tests with many nested levels of callbacks.

File: Tests/MeiliSearchUnitTests/Utils.swift
Comment (Sherlouk) [inline Tests/MeiliSearchUnitTests/Utils.swift:13]: This isn't a necessary change for the sake of async/await switching, but utilising a common extension here allows us to remove more force unwraps increasing safety throughout tests, and shorter call site increases readability elsewhere.

**Comment by curquiza** [comment]
Sorry for the delay @Sherlouk 

I'm ok to go with solution 2

Let me know when you think this is ready to be merged for you!

**Comment by Sherlouk** [comment]
Thanks Clem, will give @aronbudinszky some time to reply too to ensure we're all aligned and then will get it finished.

**Review by curquiza — APPROVED** [review]

**Comment by Sherlouk** [comment]
bors merge

**Comment by Sherlouk** [comment]
Going to send this through for now to unlock further work as this is a predecessor for quite a few bits. 

If Aron or any other consumer of the application has any opinions on how we can improve this epic then I'd love to hear it - feel free to open a new issue and we can iterate on this together!

**Pull request merged**
